### PR TITLE
Make it configurable when to exclude dangling parens for vertical

### PR DESCRIPTION
**Pull request by mads-hartmann**
The configuration defaults to the previous behavior.

### CODE REVIEW

**Comment by mads-hartmann** [comment]
By the way, this is very much a proof of concept. I'd just like to discuss the behavior for now :)

**Comment by mads-hartmann** [comment]
The previous formatting of 

```scala
case class Foo(x: String) extends Foobar
case class Bar(x: String, y: String) extends Foobar
case class VeryLongNames(xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx: String)
```

Looks like this

```scala
sealed trait Foobar
case class Foo(x: String) extends Foobar
case class Bar(
  x: String,
  y: String)
    extends Foobar
case class VeryLongNames(
  xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx: String)
```

Where the new one looks like this

```scala 
sealed trait Foobar
case class Foo(x: String) extends Foobar
case class Bar(
  x: String,
  y: String
) extends Foobar
case class VeryLongNames(
  xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx: String
)
```

**Comment by mads-hartmann** [comment]
Another interesting approach would be to say that it respects the normal `danglingParentheses` options but then not consider a paren to be dangling if it has an explicit return type or and `extends`  coming after it. 

Which would result in

```scala
sealed trait Foobar
case class Foo(x: String) extends Foobar
case class Bar(
  x: String,
  y: String
) extends Foobar
case class VeryLongNames(
  xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx: String)
```

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtConfig.scala [hunk 1]
```diff
@@ -146,6 +146,7 @@ case class ScalafmtConfig(
     poorMansTrailingCommasInConfigStyle: Boolean = false,
     verticalMultilineAtDefinitionSite: Boolean = false,
     verticalMultilineAtDefinitionSiteArityThreshold: Int = 100,
+    verticalMultilineAtDefinitionSiteDanglingExclude: List[String] = List("class", "trait"),
```
Comment (olafurpg) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtConfig.scala]: I propose we introduce a `VerticalMultiline` case class with the nested settings, similar to `Align` and `Newlines`. We can still keep the old `verticalMultilineAtDefinitionSite` `verticalMultilineAtDefinitionSiteArityThreshold` for backwards compat.

```
verticalMultiline {
  defnSite = true
  arityThreshold = 100
  ...
}
```

Comment (mads-hartmann) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtConfig.scala]: Definitely, it was getting a bit messy 👍

**Comment by pjrt** [comment]
But why? It is so ugly. You haven't seen the ugliness because the test you used were just plain case classes with no body. 

Try doing `case class X(x: Int) { def hello }` or something like that.

File: scalafmt-tests/src/test/resources/test/verticalMultilineDangling.source [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+>>>
+sealed trait Foobar
+case class Foo(x: String) extends Foobar
+case class Bar(
+  x: String,
+  y: String
+) extends Foobar
+case class VeryLongNames(
+  xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx: String
+)
```
Comment (pjrt) [inline scalafmt-tests/src/test/resources/test/verticalMultilineDangling.source]: How does it behave for `case class X()`?

**Comment by mads-hartmann** [comment]
Ugliness is relative ☺️ Do you prefer this

```scala
case class Bar(
  x: String,
  y: String)
    extends Foobar
```

Over this?

```scala
case class Bar(
  x: String,
  y: String
) extends Foobar
```

**Comment by pjrt** [comment]
I prefer the second one, which is why `verticalMultiline` does the second one. Note that the dangling does happen if u extend something. It doesn't dangle if you do not extend anything. 

That was a bad example (that's ugly regardless of how u do it). 

More lile

```
case class X(x: Int) {
   def hello
}
```

If we force dangle, then it looks like this:

```
case class X(
    x: Int
  ) {
   def hello
}
```

Which is...ugly. But yes, it is relative. I feel it looks better if the `)` is inline in that case. It is an option though...

**Comment by olafurpg** [comment]
FYI, 
```scala
case class X2(
  x: Int,
  y: Int
) { def hello: String = "Hello" }
```
becomes

```scala
case class X2(
  x: Int,
  y: Int
) {
  def hello: String = "Hello"
}
```
if you insert a line break inside the template.

**Comment by olafurpg** [comment]
One reason I'm excited about https://github.com/olafurpg/scala-syntax is that tree printing makes it easier to enforce "long-distance" consistency

**Comment by mads-hartmann** [comment]
Once #1155 is merged I'll rebase this and write some docs ☺️ Until then, here's two things to consider

- If merged this is a breaking change configuration wise
- I've deleted some of the docs from the source. IMO examples belong in the public docs. WDYT?

**Comment by mads-hartmann** [comment]
screen shot 2018-04-28 at 23 24 37

**Comment by mads-hartmann** [comment]
@olafurpg @pjrt This is now in a state where it makes it possible to

- Configure when to use dangling parens using the excludeDanglingParens option

- Opt-in to adding a newline after open parens. This is useful when formatting methods with multiple parameter lists (This fixes #1152)

Additionally this moves all vertical multiline related configuration options into a single key `verticalMultiline` - this is a breaking change configuration wise.

**Comment by olafurpg** [comment]
> Additionally this moves all vertical multiline related configuration options into a single key verticalMultiline - this is a breaking change configuration wise.


That's great! Is it possible to respect the old configuration options to avoid breaking existing .scalafmt.conf?

**Comment by mads-hartmann** [comment]
@olafurpg Should be, I was just being slightly lazy ;) I'll have a look at making it backwards compatible

**Comment by mads-hartmann** [comment]
@olafurpg Done ☺️

**Comment by mads-hartmann** [comment]
I ran it on our own codebase and there are still a couple of edge-cases I need to handle :)

**Comment by mads-hartmann** [comment]
I don't have any outstanding issues now :)

**Comment by mads-hartmann** [comment]
@olafurpg No more breaking changes and it works-as-expected-on-our-code-base™️ 😉

File: readme/Configuration.scalatex [hunk 3]
```diff
@@ -560,66 +557,104 @@
...
+      Default: @b(default.verticalMultiline.newlineAfterImplicitKW)
+
       @p If true, add a newline after an implicit keyword in function and class definitions.
 
       @fullWidthDemo(verticalAlignImplicitAfter)
         object a {
           def format(code: String, age: Int)(implicit ev: Parser, c: Context): String
         }
 
+    @sect{verticalMultiline.excludeDanglingParens}
```
[the remark this answers is not in the document]
Comment (mads-hartmann) [reply readme/Configuration.scalatex]: I was wondering what a Defn.Trait was 😅I'd definitely love to kill this config altogether. I only included it to preserve backwards compatibility - the previous implementation would accept dangling parens some places (methods) but not everywhere (constructors).

**Comment by olafurpg** [comment]
That sounds good to me @mads-hartmann

**Comment by olafurpg** [comment]
Does this PR impose changed behavior from how people used vertical multiline before? I just unblocked #1145 and I'd like to rebase on top of the changes here to clean up the remaining usage of `toHocon`

**Comment by mads-hartmann** [comment]
@olafurpg In its current state it does not :) However if we remove the `excludeDanglingParens` option in favour of relying on `danglingParens` it will as the previous behaviour would add a dangling paren for methods but not case-class constructors - using `danglingParens` it wouldn't be able to differentiate between the two :)

**Comment by mads-hartmann** [comment]
@olafurpg Awesome, thanks 👍

**Pull request merged**